kafka/consumer.go

```diff
@@ -249,3 +239,165 @@ func (c *Consumer) Healthy() error {
 	}
 	return nil
 }
+
+// consumer wraps partitionConsumers and exposes the necessary callbacks
+// to use when partitions are reassigned.
+type consumer struct {
+	mu        sync.Mutex
+	consumers map[topicPartition]partitionConsumer
+	processor model.BatchProcessor
+	logger    *zap.Logger
+	decoder   Decoder
+	delivery  apmqueue.DeliveryType
+}
+
+type topicPartition struct {
+	topic     string
+	partition int32
+}
+
+// assigned must be set as a kgo.OnPartitionsAssigned callback. Ensuring all
+// assigned partitions to this consumer process received records.
+func (c *consumer) assigned(_ context.Context, client *kgo.Client, assigned map[string][]int32) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	for topic, partitions := range assigned {
+		for _, partition := range partitions {
+			pc := partitionConsumer{
+				stop:      make(chan struct{}),
+				done:      make(chan struct{}),
+				records:   make(chan []*kgo.Record),
+				processor: c.processor,
+				logger:    c.logger,
+				decoder:   c.decoder,
+				client:    client,
+				delivery:  c.delivery,
+			}
+			go pc.consume(topic, partition)
+			tp := topicPartition{partition: partition, topic: topic}
+			c.consumers[tp] = pc
+		}
+	}
+}
+
+// lost must be set as a kgo.OnPartitionsLost and kgo.OnPartitionsReassigned
+// callbacks. Ensures that partitions that are lost (see kgo.OnPartitionsLost
+// for more details) or reassigned (see kgo.OnPartitionsReassigned for more
+// details) have their partition consumer stopped.
+// This callback must finish within the re-balance timeout.
+func (c *consumer) lost(_ context.Context, _ *kgo.Client, lost map[string][]int32) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	var wg sync.WaitGroup
+	for topic, partitions := range lost {
+		for _, partition := range partitions {
+			tp := topicPartition{topic: topic, partition: partition}
+			pc := c.consumers[tp]
+			delete(c.consumers, tp)
+			close(pc.stop)
+			wg.Add(1)
+			go func() { <-pc.done; wg.Done() }()
+		}
+	}
+	wg.Wait()
```

Comment:
I was reading the code again and I don't think it's necessary since at this point the polling is stopped (rebalance is happening), and once the entry is removed from the map, it isn't accessible by TopicPartition anymore.